src/components/assetlib_projects/assetlib_project_saving_dialog.gd

```diff
@@ -0,0 +1,127 @@
+		if _project_downloader.get_http_client_status() == HTTPClient.STATUS_BODY:
+			if _project_downloader.get_body_size() > 0:
+				_message_label.text = "%s (%s / %s)..." % [
+					tr("Downloading"),
+					String.humanize_size(_project_downloader.get_downloaded_bytes()),
+					String.humanize_size(_project_downloader.get_body_size())
+				]
+			else:
+				_message_label.text = "%s (%s)..." % [
+					tr("Downloading"),
+					String.humanize_size(_project_downloader.get_downloaded_bytes())
+				]
+		if _project_downloader.get_http_client_status() == HTTPClient.STATUS_RESOLVING:
+			_message_label.text = tr("Resolving...")
+		elif _project_downloader.get_http_client_status() == HTTPClient.STATUS_CONNECTING:
+			_message_label.text = tr("Connecting...")
+		elif _project_downloader.get_http_client_status() == HTTPClient.STATUS_REQUESTING:
+			_message_label.text = tr("Requesting...")
+		await get_tree().create_timer(0.1).timeout
+
+
+func _on_project_downloader_request_completed(result: int,
+		response_code: int, _headers: PackedStringArray,
+		_body: PackedByteArray):
+	if result != HTTPRequest.RESULT_SUCCESS or response_code != 200:
+		_error(tr("Download failed."))
+		get_ok_button().disabled = false
+		_download_cleanup()
+		return
+	
+	var dir = _project_path_line_edit.text.strip_edges()
+	var project_name = _project_name_edit.text.strip_edges()
+	_ZIP.unzip(_project_downloader.download_file, dir.get_base_dir())
+	
+	var zip = ZIPReader.new()
+	zip.open(_project_downloader.download_file)
+	var err = _unzip_to_path(zip, dir)
+	if err != OK:
+		assert(err != ERR_FILE_NOT_FOUND)
+		_error(error_string(err))
+		get_ok_button().disabled = false
+		_download_cleanup()
+		return
+
+	var cfg = ConfigFile.new()
+	err = cfg.load(dir.path_join("project.godot"))
+	if err != OK:
+		_error(error_string(err))
+		get_ok_button().disabled = false
+		_download_cleanup()
+		return
+	cfg.set_value("application", "config/name", project_name)
+	err = cfg.save(dir.path_join("project.godot"))
+	if err != OK:
+		_error(error_string(err))
+		get_ok_button().disabled = false
+		_download_cleanup()
+		return
+	
+	_success("Download completed.")
+	_download_cleanup()
+	created.emit(dir)
+
+
+## A procedure that unzips a zip file to a target directory, keeping the
+## target directory as root, rather than the zip's root directory.
+func _unzip_to_path(zip: ZIPReader, destiny: String) -> Error:
+	var files = zip.get_files()
+	var err
+	
+	for zip_file_name in files:
+		if zip_file_name == files[0]:
+			continue
+		var target_file_name = destiny.path_join(zip_file_name.split("/", false, 1)[1])
+		if zip_file_name.ends_with("/"):
+			err = DirAccess.make_dir_recursive_absolute(target_file_name)
+			if err != OK:
+				return err
+		else:
+			var file_contents = zip.read_file(zip_file_name)
+			var file = FileAccess.open(target_file_name, FileAccess.WRITE)
+			if not file:
+				return FileAccess.get_open_error()
+			file.store_buffer(file_contents)
+			file.close()
+	return OK
```

Comment:
> By the way, why does extensions/zip.gd use external programs, rather than the builtin ZipReader class?
because of how apps are handled in macos. We are able to execute a downloaded editor only if it was unzipped with os.execute(unzip, ...) (at least it is the only way I've found so far)